File: script/prove_batch.py
#!/usr/bin/env python3
"""
Batch proof oracle for randomized differential testing.

Usage:
    prove_batch.py <count> <seed> [maxlen]

Generates `count` random lengths in [0, maxlen], builds a sound proof for each
(cross-checked against the official `blake3` package when available), and prints
the ABI-encoding of parallel arrays:

    (uint64[] lengths, bytes32[] roots, bytes[] finalChunks, bytes32[][] edges)

as a single 0x-hex string. Amortizes interpreter startup across many vectors.
"""
import random
import sys

from eth_abi import encode

import blake3_tree as b3


def main():
    count = int(sys.argv[1])
    seed = int(sys.argv[2])
    maxlen = int(sys.argv[3]) if len(sys.argv) > 3 else 70000

    rng = random.Random(seed)
    try:
        import blake3
        have_oracle = True
    except ImportError:
        have_oracle = False

    lengths, roots, finals, edges = [], [], [], []
    for _ in range(count):
        L = rng.randint(0, maxlen)
        data = rng.randbytes(L)
        root, alen, fc, edge = b3.make_sound_proof(data)
        if have_oracle:
            assert root == blake3.blake3(data).digest()
        lengths.append(alen)
        roots.append(root)
        finals.append(fc)
        edges.append(edge)

    enc = encode(
        ["uint64[]", "bytes32[]", "bytes[]", "bytes32[][]"],
        [lengths, roots, finals, edges],
    )
    sys.stdout.write("0x" + enc.hex())


if __name__ == "__main__":
    main()

File: prover/src/bin/prove_batch.rs
//! prove-batch <count> <seed> [maxlen]
//!
//! Prints the ABI-encoding of parallel arrays
//! (uint64[] lengths, bytes32[] roots, bytes[] finalChunks, bytes32[][] edges)
//! for randomized differential testing.

use bao_length_prover::{make_data, make_proof};
use ethabi::{encode, Token};
use ethereum_types::U256;

// Deterministic length picker (SplitMix64), independent of data filler.
struct Rng(u64);
impl Rng {
    fn next(&mut self) -> u64 {
        self.0 = self.0.wrapping_add(0x9E3779B97F4A7C15);
        let mut z = self.0;
        z = (z ^ (z >> 30)).wrapping_mul(0xBF58476D1CE4E5B9);
        z = (z ^ (z >> 27)).wrapping_mul(0x94D049BB133111EB);
        z ^ (z >> 31)
    }
}

fn main() {
    let args: Vec<String> = std::env::args().collect();
    let count: usize = args[1].parse().expect("count");
    let seed: u64 = args[2].parse().expect("seed");
    let maxlen: usize = args.get(3).map(|s| s.parse().unwrap()).unwrap_or(70000);

    let mut rng = Rng(seed);
    let (mut lens, mut roots, mut finals, mut edges) = (vec![], vec![], vec![], vec![]);
    for i in 0..count {
        let length = (rng.next() as usize) % (maxlen + 1);
        let data = make_data(length, seed ^ (i as u64));
        let p = make_proof(&data);
        lens.push(Token::Uint(U256::from(p.length)));
        roots.push(Token::FixedBytes(p.root.to_vec()));
        finals.push(Token::Bytes(p.final_chunk));
        edges.push(Token::Array(
            p.right_edge.iter().map(|cv| Token::FixedBytes(cv.to_vec())).collect(),
        ));
    }

    let enc = encode(&[
        Token::Array(lens),
        Token::Array(roots),
        Token::Array(finals),
        Token::Array(edges),
    ]);
    print!("0x{}", hex_encode(&enc));
}

fn hex_encode(bytes: &[u8]) -> String {
    let mut s = String::with_capacity(bytes.len() * 2);
    for b in bytes {
        s.push_str(&format!("{:02x}", b));
    }
    s
}
